### PR TITLE
Implement Java interface native OS library name generation using Maven

### PR DESCRIPTION
**Pull request by marko-asplund**
This PR aims to resolve issues reported in #780 and #807 by implementing native OS library name generation during build-time using Maven.

@jmorra and @petro-rudenko: could you please review and test.

This PR depends on PR #821.

Also, created an issue for `os-maven-plugin` to provide additional OS info: https://github.com/trustin/os-maven-plugin/issues/15

### CODE REVIEW

**Comment by jmorra** [comment]
This doesn't build properly on a mac.  When I run I see the following snippets

```
[INFO] ------------------------------------------------------------------------
[INFO] Detecting the operating system and CPU architecture
[INFO] ------------------------------------------------------------------------
[INFO] os.detected.name: osx
[INFO] os.detected.arch: x86_64
[INFO] os.detected.classifier: osx-x86_64
[INFO]                                         
```

and

```
[INFO] --- copy-rename-maven-plugin:1.0:copy (copy-file) @ vw-jni ---
[INFO] Copied /Users/jmorra/git/vowpal_wabbit/java/target/vw_jni.lib to /Users/jmorra/git/vowpal_wabbit/java/target/classes/vw_jni-${os.detected.release}_${os.detected.release.version}_x86_64.lib
```

The file created on my machine is called `vw_jni-${os.detected.release}_${os.detected.release.version}_x86_64.lib	vw_jni-osx_${os.detected.release.version}_x86_64.lib` which isn't correct.

File: java/src/main/java/vw/jni/NativeUtils.java [hunk 1]
```diff
@@ -49,7 +49,7 @@ public static String getLinuxVersion() throws IOException {
             Process process = Runtime.getRuntime().exec("lsb_release -r");
             reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
             String line;
-            Pattern releasePattern = Pattern.compile("Release:\\s*(\\d+).*");
+            Pattern releasePattern = Pattern.compile("Release:\\s*(\\S+).*");
```
Comment (jmorra) [inline java/src/main/java/vw/jni/NativeUtils.java]: Why this change?  This regex is slower.

File: java/src/main/java/vw/jni/NativeUtils.java [hunk 2]
```diff
@@ -88,7 +88,7 @@ else if (osName.toLowerCase().contains("linux")) {
             if (version == null) {
                 throw new UnsupportedOperationException("Cannot determine linux version");
             }
-            return distro.trim().replaceAll(" ", "_") + "." + version;
+            return (distro.trim().replaceAll(" ", "_") + "_" + version).toLowerCase();
```
Comment (jmorra) [inline java/src/main/java/vw/jni/NativeUtils.java]: If you change the pattern here you have to change `build.sh` as well.  If you don't update this file at all then this change can be merged in without affecting `build.sh`.  I had intended the `.` to separate out different elements of the build environment in the name and `_` to denote word separation.  For example, in my opinion `vw_jni.Red_Hat.6.amd64.lib` is more legible than `vw_jni.Red_Hat_6_amd64.lib`.  As far as casing, this is designed to match the system calls to uname.  If you look at line 171 of `build.sh` you can see I specifically don't input `Darwin` but instead get the property directly.  I think this is more robust as I'm just reusing the system properties.

File: java/src/main/java/vw/jni/NativeUtils.java [hunk 3]
```diff
@@ -103,7 +103,7 @@ else if (osName.toLowerCase().contains("linux")) {
      */
     public static void loadOSDependentLibrary(String path, String suffix) throws IOException {
         String osFamily = getOsFamily();
-        String osDependentLib = path + "." + osFamily + "." + System.getProperty("os.arch") + suffix;
```
Comment (jmorra) [inline java/src/main/java/vw/jni/NativeUtils.java]: See above.

File: java/src/main/java/vw/jni/NativeUtils.java (hunk 1 above)
Comment (marko-asplund) [inline java/src/main/java/vw/jni/NativeUtils.java]: The original regex parses OS release major version, but unfortunately `os-maven-plugin` only supports getting the full version.

Comment (jmorra) [inline java/src/main/java/vw/jni/NativeUtils.java]: This is problematic then because if we go with this we'll have to create a separate build for every minor release as well and the number of libraries packaged in the jar will explode.

File: java/src/main/java/vw/jni/NativeUtils.java (hunk 2 above)
Comment (marko-asplund) [inline java/src/main/java/vw/jni/NativeUtils.java]: The rationale behind proposing a naming scheme change was that the syntax would allow easier parsing of different name components. Consider the following example based on the original scheme:
`vw_jni.ubuntu.13.10.amd64.lib`
arguably this would be somewhat trickier to parse than 
`vw_jni-ubuntu_13.10_amd64.lib`

So, that was the rationale, but I don't have a problem with using the original scheme either, if you think that it would be better.

Also, `os-maven-plugin` lowercases strings, so I think we should follow that convention.

Comment (jmorra) [inline java/src/main/java/vw/jni/NativeUtils.java]: Lowercasing should be fine, but we have to update `build.sh` then.  It should be straightforward to do, just update lines 19, 24, 29, 49, 56, 61, and 171.  The last line might be a little tricky.  I really think adding the minor version to the name is problematic.

Comment (marko-asplund) [inline java/src/main/java/vw/jni/NativeUtils.java]: Understand your concern about major/minor OS version. 
Using the `os-maven-plugin` based approach, one way around this would be to implement  https://github.com/trustin/os-maven-plugin/issues/15.
Another option would be to list all the native libraries included in the archive at runtime and then automatically pick the best match.
Not sure which would be the better approach.
With Linux, binary compatibility it's not so much about the distribution and version as much as it is about glibc and kernel version. In that respect, the latter approach could arguably be more robust.

**Comment by JohnLangford** [comment]
It sounds like build.sh needs to be fixed here.   

I'm still sketchy on why the change of naming is worthwhile?  What is the scenario which calls for this?

**Comment by marko-asplund** [comment]
Here's some prototype code that would allow listing available native library files:
https://gist.github.com/marko-asplund/58f91f386c363e89fd3d

`NativeUtils.loadOSDependentLibrary` could use something like this to iterate over available library files and pick the most suitable one. I think this could be one way of getting over the major/minor issue.

However, I think determining binary compatibility is a bit problematic. OS type and CPU architecture most likely need to match, but e.g. different Linux variants can still be compatible. Should the OS release and version be included in the native library name at all?

@jmorra what do you think?

**Comment by JohnLangford** [comment]
@jmorra  can you take a look?

My understanding is that the core problem is linux likes to name itself in many ways, so using the right flavor of binary is tricky.

**Comment by jmorra** [comment]
That's exactly right. Not only that, but this doesn't work on Mac. I would
hold off on this until the referenced issue is resolved with the maven
plugin.

**Comment by marko-asplund** [comment]
Yes, this still requires some more work, so it's not ready to be merged.
I created PR https://github.com/trustin/os-maven-plugin/pull/16 for `os-maven-plugin` to address some requirements of this PR, but that hasn't been processed yet.

**Comment by JohnLangford** [comment]
I'm going to close this for now.  Please reopen when you fix the issues.
